lib/logstash/codecs/csv.rb

```diff
@@ -88,6 +92,7 @@ class LogStash::Codecs::CSV < LogStash::Codecs::Base
   CONVERTERS.freeze
 
   def register
+    @buffer = FileWatch::BufferedTokenizer.new(@delimiter)
```

Comment:
Using the buffered tokenizer is tricky here, because CSV data can have literal newlines as long as they are quoted. By pre-tokenizing before handing off to the CSV parser that has the context, we effectively prevent _legal_ literal newlines from being used.

Additionally, input plugins that already use the `BufferedTokenizer` (e.g. File Input) will _strip_ the newlines from their input before passing off each entry to the codec.

I think another approach would be to use `CSV::parse` instead of `CSV::parse_line`, and then iterate over the resulting entries.

Reply:
TIL line breaks inside CSV values 😮

Reply:
CSV in practice is a giant mess, because it developed organically without a formal specification over the course of decades. When one product encountered an edge-case, they came up with a solution to their problem but often solved it in a way that introduced new and weirder edge cases (e.g., with no agreed-upon escape sequence, some solved the comma- and newline-in-field problem by adding quoting, which made quote characters magical and precludes our ability to pre-tokenize records)

Ruby's CSV implementation is pretty robust to the variety of data under the CSV umbrella, giving us options like `row_sep` to control the record delimiter, `col_sep` to control the field delimiter, `quote_char` to control how it understands quoted sequences, etc.

I included a recommendation in my previous review to use `CSV::parse`, instead of `CSV::parse_line`, because it is capable of handling multiple entries but otherwise remains the same (the `parse_line` variant simply ignores any additional records). We can still pipe through the "delimiter" option to CSV's `row_sep` parameter, and it should handle quoted literal row separators for us..

Reply:
Yes, I did test `CSV::parse` and it does work correctly in handling both line breaks in columns and line breaks at end of line but it does not work for *streaming* input scenarios where the `BufferedTokenizer` is useful but breaks the line breaks in columns case. Will recap and continue discussion in main thread.